### ISSUE TITLE
Regex for failure items fails if the junit xml is without colors

### ISSUE THREAD

**Issue by MajorDallas** [CarloDePieri/vim-pytest#2]
With no commits in over 2 years, I'm hoping this isn't just a shout in the wind, but so far this is the only pytest integration I've found that does what I want.

Unfortunately my first attempt at running it failed, and I was able to track the problem to the regex defined on line 67 in lib.py:

```python
r = r"#x1B\[1m#x1B\[31m(.*)#x1B\[0m:(\d*): (.*)"
```

By disabling the exception handler I found that the `re.search` call on line 98 was returning a `None` due to no matches.

I tried wrapping the control codes in a non-capturing group with the `?` quantifier and was able to resolve the error:

```python
r = r"(?:#x1B\[1m#x1B\[31m)?(.*)(?:#x1B\[0m)?:(\d*): (.*)"
```

This likely needs to be applied to all the regexs in the script, but at least with my current test suite I haven't tripped over those yet.

**Comment by MajorDallas** [comment]
Also, the regex `file\ (.*), line\ (\d*)` seems to no longer be applicable to how pytest ouputs this information. I've had success changing it to `([^:\n]*):(\d*):[ \t]*([^\n]*)`. (I started with a much more specific regex, but it locked up my laptop :sweat_smile: )

**Comment by CarloDePieri** [comment]
I'm having trouble confirming the first issue, that regex seems to be working on my tests! Let's look into it: try to set a specific xml file by putting somewhere in your config:
```
let g:pytest_xml_file="/tmp/test-output.xml" 
```
Then launch a minimal broken test and copy-paste the resulting xml file here.
Also, which version of python / pytest are you running?

Regarding the regex catering to errors which are not "collection error'', you are correct. A better regexp for that line seems to be `#x1B\[1m#x1B\[31m(.*)#x1B\[0m\:(\d*)\:`. I'll fix this as soon as we figure out the first issue.

**Comment by MajorDallas** [comment]
Python 3.9.16
Pytest 4.6.11

```ini
# pytest.ini
[pytest]
junit_family = xunit2
```

Quickfix entry:
```
test_vimpytest.py|1 error| Execution error. Check :PytestOutput
```

`:PytestOutput` is a perfectly normal Pytest report:

```
============================= test session starts ==============================
platform linux -- Python 3.9.16, pytest-4.6.11, py-1.11.0, pluggy-0.13.1
rootdir: ...., inifile: pytest.ini
plugins: assume-2.4.3, anyio-3.7.1
collected 1 item

test_vimpytest.py F                                                      [100%]

=================================== FAILURES ===================================
__________________________________ test_fail ___________________________________

    def test_fail():
>       assert False
E       assert False

test_vimpytest.py:4: AssertionError
-------------------- generated xml file: /tmp/nvimXCYWyp/3 ---------------------
=========================== 1 failed in 0.06 seconds ===========================

```

```python
# test_vimpytest.py
import pytest

def test_fail():
    assert False
```

And the xml file:
```xml
<?xml version="1.0" encoding="utf-8"?><testsuites><testsuite errors="0" failures="1" hostname="..." name="pytest" skipped="0" tests="1" time="0.061" timestamp="2023-08-04T17:58:55.701026"><testcase classname="tests.test_vimpytest" name="test_fail" time="0.001"><failure message="assert False">def test_fail():
&gt;       assert False
E       assert False

test_vimpytest.py:4: AssertionError</failure></testcase></testsuite></testsuites>
```


After adding `?` and non-capture groups to that regex, my quickfix list shows

```
test_vimpytest.py|4 F| AssertionError: assert False        (test_fail)
```

**Comment by CarloDePieri** [comment]
Oh, you have no color _in the xml_, that's why it's failing. I misunderstood the issue title, I thought it was referring to your terminal.

Just out of curiosity, are you manually disabling them with `PY_COLORS=0` or something similar? Because even `--color=no` leaves colors enabled in the xml output, that's why I had not considered to support xmls without colors. But I guess it would make sense to support them.

ANSI colors codes here are handy and precise delimiters for the regexes. Without them, the start of the line would be a good candidate for a delimiter (simply not capturing colors won't work; your proposed regex would catch a comment like `# The meaning of life? :42: of course!`). Let me take a stab at this.

**Comment by MajorDallas** [comment]
I'm not _manually_ disabling anything afaik. It could be a default on Fedora, or set by Poetry, a quirk of this version of Pytest... Tbh I was kinda thinking you were manually _enabling_ colors :sweat_smile: 

Point taken about them being a handy way to spot the info of interest, though, I hadn't thought of that! Maybe a better approach is to force colors to be on, rather than find the perfect regex that gets exactly and only what we want?

Update: I'm not setting `PY_COLORS=0` anywhere, but I just tested with running `export PY_COLORS=1` before starting nvim and the resulting XML now has ANSI control codes in it; somehow using pytest's own `--color` option had no effect one way or another. 

After pulling 7fdf321, I'm getting `WrongRegexException` on a "test setup failure". Adding a couple `print`s, I got:

```python
if not group(name):
    print(name)  # message_c
    print(found.groups())  # ('../tests/conftest.py', '354', '', None, None, None)
    print(found.groupdict()  # {'file_c': '../tests/conftest.py', 'line_c': '354', 'message_c': '', 'file_n': None, 'line_n': None, 'message_n': None}
    print(found)  # <re.Match object; span=(974, 1022), match='#x1B[1m#x1B[31m../tests/conftest.py#x1B[0m:354: '>
```

Notably, the matched string does not contain `^E\ \ \ `... This particular error has multiple stacktraces in it, separated by the same delimiter as failures with multiple stacktraces. In this case, the "E" string is in the second stacktrace and looks like

```
#x1B[1m#x1B[31mE               asks.errors.BadStatus: 403 Client Error: Forbidden for url: ....
```

but it's not being captured because `#x1B[1m#x1B[31m../tests/conftest.py#x1B[0m:354:` appears first and the `WrongRegexException` is raised for it before this "E" string can be matched.

Full disclosure: I'm probably abusing fixtures in order to get this. My only defense is I didn't what I was doing when I wrote the test suite and I'm aware it's time for some rewrites :sweat_smile: . That said, the format for errors and failures when multiple stacktraces are present doesn't _look_ that different at a glance...

Also want to add that I really appreciate you taking the time to look into this :smile: I can imagine seeing a new issue notification was a surprise and that you have plenty of other things to do. So, thank you, and lmk what else I can do to help.

**Comment by CarloDePieri** [comment]
Testing the plugin on code bases other than mines is essential to find out possible edge cases, so you are more than welcome with these reports!

I added the `WrongRegexException` because I wanted a clear warning when a match was not found despite me expecting one: either the code is wrong (possible :sweat_smile: ) or this is an edge case I didn't anticipate; it would be nice to support it if possible, even if it's remote.

This is clearly even not that remote, a fixture that calls a normal function that fails causes a layered stacktrace and that's not handled right now: I'll get to it.

What you can further do to help, if you find other instances like this, is to write and post the minimal test that reproduce the issue (right now it's not needed).

**Comment by CarloDePieri** [comment]
@MajorDallas try and see if [bb36f8a](https://github.com/CarloDePieri/vim-pytest/commit/bb36f8a486c9aedf08620aca3a28507cc8f7263f) is now working for you!

**Comment by MajorDallas** [comment]
Works perfectly! It's putting into quickfix both the fixture that failed and the code it called into that failed--even better than I hoped for :smile: Thank you!

**Comment by CarloDePieri** [comment]
Seems to be working as intended then!
I'm going ahead and merge this branch, if something else come up feel free to open another issue :stuck_out_tongue_winking_eye:
